### PR TITLE
feat: Drop jQuery requirement 

### PR DESCRIPTION
**Pull request by MarkoBL**
Sorry, I messed it up to change the target (I accidentally renamed the branch). This closed the original pull request: https://github.com/6pac/SlickGrid/pull/719

### CODE REVIEW

File: slick.core.js [hunk 1]
```diff
@@ -38,6 +48,8 @@
      */
     this.stopImmediatePropagation = function () {
       isImmediatePropagationStopped = true;
+      if(nativeEvent)
+        nativeEvent.stopImmediatePropagation();
```
Comment (ghiscoding) [inline slick.core.js:52]: could you wrap this `if` and the other function `if` as well in curly brace to avoid potential issues in the future

File: slick.core.js [hunk 2]
```diff
@@ -48,6 +60,36 @@
...
+    }
+
+    this.preventDefault = function() {
+      if(nativeEvent)
+        nativeEvent.preventDefault();
+      isDefaultPrevented = true;
+    }
+
+    this.isDefaultPrevented = function() {
+      if(nativeEvent)
```
Comment (ghiscoding) [inline slick.core.js:75]: same comment as above to wrap in curly braces to avoid future bugs, there are a few of these in this file, they could be all be changed at the same time please

File: slick.core.js [hunk 3]
```diff
@@ -96,15 +138,17 @@
...
 
-      var returnValue;
       for (var i = 0; i < handlers.length && !(e.isPropagationStopped() || e.isImmediatePropagationStopped()); i++) {
-        returnValue = handlers[i].call(scope, e, args);
+        const returnValue = handlers[i].call(scope, e, args);
+        e.addReturnValue(returnValue);
       }
 
-      return returnValue;
+      return e;
```
Comment (ghiscoding) [inline slick.core.js:151]: if I wonder if we could simplify this to return an object as `{ event, value }` instead of having to call `getReturnValue()`, this way we would get both in the same return. It could also be changed in a future PR

File: slick.core.js [hunk 4]
```diff
@@ -675,10 +721,239 @@
...
+    if (value === undefined) {
+      return el.getBoundingClientRect().height;
+    }
+    setStyleSize(el, "height", value);
+  }
+
+  function setStyleSize(el, style, val) {
+    if (typeof val === 'function') val = val();
+    if (typeof val === 'string') el.style[style] = val;
+    else el.style[style] = val + 'px';
```
Comment (ghiscoding) [inline slick.core.js:772]: I'm a big fan of wrapping everything in curly braces to avoid potential future bugs, it would also make the code clearer and easier to read

File: slick.core.js [hunk 5]
```diff
@@ -675,10 +721,239 @@
...
+  }
+
+  function contains(parent, child) {
+    if(!parent || !child)
+      return false;
+
+    const parentList = parents(child);
+    return !parentList.every((p) => {
```
Comment (ghiscoding) [inline slick.core.js:780]: hmm up to this point we were still supporting the ES5 syntax and not yet syntax like `() => {}`, we might need to change this to keep compatibility since we aren't transpiling our code. Note that there a few more of these in this file too

File: slick.grid.js [hunk 6]
```diff
@@ -534,14 +543,19 @@ if (typeof Slick === "undefined") {
...
-        disableSelection($headers); // disable all text selection in header (including input and textarea)
+        disableSelection(_headers); // disable all text selection in header (including input and textarea)
 
         if (!options.enableTextSelectionOnCells) {
           // disable text selection in grid cells except in input and textarea elements
           // (this is IE-specific, because selectstart event will only fire in IE)
-          $viewport.on("selectstart.ui", function (event) {
-            return $(event.target).is("input,textarea");
-          });
+          _viewport.forEach((view) => {
```
Comment (ghiscoding) [inline slick.grid.js:551]: same as earlier comment, can we keep an ES5 syntax `function () {}` since we aren't transpiling our code, there a few of these in this file, they could all be changed at same time

File: slick.grid.js [hunk 7]
```diff
@@ -558,69 +572,67 @@ if (typeof Slick === "undefined") {
...
+        _headerRowScroller.forEach((scroller) => {
+           scroller.addEventListener("scroll", handleHeaderRowScroll);
+        });
 
         if (options.showHeaderRow) {
-          $headerRow
-            .on("mouseenter", ".slick-headerrow-column", handleHeaderRowMouseEnter)
-            .on("mouseleave", ".slick-headerrow-column", handleHeaderRowMouseLeave);
+          _headerRows.forEach((row) => {
+            //row.addEventListener("mouseover", handleHeaderRowMouseOver);
```
Comment (ghiscoding) [inline slick.grid.js:601]: what's the reason that these were commented out and removed? That is probably why you had to fix the Cypress tests on the CustomTooltip Plugin, isn't it?

File: slick.grid.js [hunk 8]
```diff
@@ -710,44 +726,44 @@ if (typeof Slick === "undefined") {
...
     }
 
-    function getActiveCanvasNode(element) {
-      setActiveCanvasNode(element);
+    function getActiveCanvasNode(e) {
 
-      return $activeCanvasNode[0];
-    }
+      if(e === undefined)
+        return _activeCanvasNode;
```
Comment (ghiscoding) [inline slick.grid.js:732]: wrap in curly brace please :) ... or return on a single line, that could also be acceptable. There are a few of these in this file

File: slick.grid.js [hunk 9]
```diff
@@ -161,32 +157,36 @@ if (typeof Slick === "undefined") {
     var vScrollDir = 1;
 
     // private
+    const u = Slick.Utils;
```
Comment (ghiscoding) [inline slick.grid.js:160]: same as earlier comments, I would prefer to see `utils` as the variable name, if it doesn't conflict with anything else that is

File: slick.grid.js [hunk 10]
```diff
@@ -4841,12 +4852,19 @@ if (typeof Slick === "undefined") {
       }
     }
 
-    function handleClick(e) {
+    function handleClick(evt) {
+
+      let e = evt;
+      if(e instanceof Slick.EventData)
+        e = evt.getNativeEvent();
```
Comment (ghiscoding) [inline slick.grid.js:4859]: wrap in curly brace please :)

File: slick.grid.js [hunk 11]
```diff
@@ -5004,54 +5036,76 @@ if (typeof Slick === "undefined") {
...
     function getCellFromEvent(e) {
+      if(e instanceof Slick.EventData)
+        e = e.getNativeEvent();
+
       var targetEvent = e.touches ? e.touches[0] : e;
       var row, cell;
-      var $cell = $(e.target).closest(".slick-cell", $canvas);
-      if (!$cell.length) {
+
+      var celll = e.target.closest(".slick-cell");
```
Comment (ghiscoding) [inline slick.grid.js:5087]: is that a typo? 3 "l" on "celll"

File: slick.grid.js [hunk 12]
```diff
@@ -6262,11 +6324,9 @@ if (typeof Slick === "undefined") {
       "getCanvasWidth": getCanvasWidth,
       "getCanvases": getCanvases,
       "getActiveCanvasNode": getActiveCanvasNode,
-      "setActiveCanvasNode": setActiveCanvasNode,
       "getViewportNode": getViewportNode,
       "getViewports": getViewports,
       "getActiveViewportNode": getActiveViewportNode,
-      "setActiveViewportNode": setActiveViewportNode,
```
Comment (ghiscoding) [inline slick.grid.js:6269]: isn't there a way to keep all existing functions? That could be a breaking for a few users, probably me included with my libs, so I would rather keep these 2 functions if possible

**Review by ghiscoding — CHANGES_REQUESTED** [review]
great start, I didn't test with my libs yet but that should be soon

**Comment by 6pac** [comment]
BTW @MarkoBL and @ghiscoding  I just patched for https://github.com/6pac/SlickGrid/issues/727 - you might want to integrate it.
It's a one-liner but was affecting performance of some frozen row/col unit tests quite severely.

Also, I see you're running across some issues with the grid returning jQuery objects from function calls, eg. for page elements.  Things like this are not only breaking changes but essentially a paradigm shift. 
We really need a good migration HOWTO summarising all the changes the developers (consumers of the grid) will have to make, probably with examples.

File: slick.editors.js [hunk 13]
```diff
@@ -4,24 +4,26 @@
  * @namespace Slick
  */
 
-(function ($) {
+(function () {
+
+  const u = Slick.Utils;
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply slick.editors.js]: Done

File: slick.editors.js [hunk 14]
```diff
@@ -4,24 +4,26 @@
...
-        .select();
+      input = u.template("<INPUT type=text class='editor-text' />", args.container);
+      input.addEventListener("keydown.nav", navOnLR ? handleKeydownLRNav : handleKeydownLRNoNav);
+      input.focus();
+      input.select();
 
       // don't show Save/Cancel when it's a Composite Editor and also trigger a onCompositeEditorChange event when input changes
       if (args.compositeEditorOptions) {
-        $input.on("change", function () {
+        input.addEventListener("change", function () {
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply slick.editors.js]: Done

File: slick.core.js (hunk 1 above)
Comment (MarkoBL) [reply slick.core.js:53]: Done

File: slick.core.js (hunk 4 above)
Comment (MarkoBL) [reply slick.core.js]: Done

File: slick.grid.js (hunk 8 above)
Comment (MarkoBL) [reply slick.grid.js:732]: Done

File: slick.grid.js (hunk 9 above)
Comment (MarkoBL) [reply slick.grid.js]: Done

File: slick.grid.js (hunk 11 above)
Comment (MarkoBL) [reply slick.grid.js]: Done

File: slick.groupitemmetadataprovider.js [hunk 15]
```diff
@@ -14,6 +14,7 @@
    * @param inputOptions
    */
   function GroupItemMetadataProvider(inputOptions) {
+    const u = Slick.Utils;
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply slick.groupitemmetadataprovider.js]: Done

File: slick.editors.js [hunk 16]
```diff
@@ -4,24 +4,26 @@
...
     this.args = args;
 
     this.init = function () {
       var navOnLR = args.grid.getOptions().editorCellNavOnLRKeys;
-      $input = $("<INPUT type=text class='editor-text' />")
-        .appendTo(args.container)
-        .on("keydown.nav", navOnLR ? handleKeydownLRNav : handleKeydownLRNoNav)
-        .focus()
-        .select();
+      input = u.template("<INPUT type=text class='editor-text' />", args.container);
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply slick.editors.js]: Renamed all upper case elements in this file.

File: slick.grid.js (hunk 10 above)
Comment (MarkoBL) [reply slick.grid.js:4865]: Done

File: slick.grid.js (hunk 7 above)
Comment (MarkoBL) [reply slick.grid.js:601]: Well, I'm not a web guy, therefore I always learn new stuff whenever I do some javascript :) And I learned that javascript has two different ways to trigger a hover event: mouseover/out and mouseenter/leave. And I actually don't know what jQuery is doing under the hood, therefore I was playing around a little bit and probably forgot to remove this old code.

Anyway, the canvas uses mouseover/mouseout and all the other elements use mouseenter/mouseleave.

The problem with your cypress test was, that you mixed mouseover/mouseleave (or the other way around, I don't know anymore). While this works with the secret jQuery sauce and all the events where called, it didn't work in pure javascript. That made me mad and took some time to debug :)

File: slick.core.js (hunk 5 above)
Comment (MarkoBL) [reply slick.core.js]: Done

File: slick.grid.js (hunk 6 above)
Comment (MarkoBL) [reply slick.grid.js]: Done

File: examples/example1-simple.html [hunk 17]
```diff
@@ -26,8 +26,6 @@ <h2>View Source:</h2>
   </tr>
 </table>
 
-<script src="../lib/jquery-3.1.0.js"></script>
-<script src="../slick.interactions.js"></script>
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply examples/example1-simple.html:30]: It was included two times in the original version, look at here: https://github.com/6pac/SlickGrid/blob/008ed8886f298fe7dbd64c0c890d8cbdc2eb3b87/examples/example1-simple.html#L34

File: examples/example-composite-editor-item-details.html [hunk 18]
```diff
@@ -178,7 +178,7 @@ <h2>View Source:</h2>
 
 
     var containers = $.map(columns, function (c) {
```
[the remark this answers is not in the document]
Comment (MarkoBL) [reply examples/example-composite-editor-item-details.html:180]: Well, it's not prohibited to use SlickGrid and jQuery together :) There are 77 examples included, and I don't see the use to modify every example. Maybe some important ones, but not all. And the plugins still make use of jQuery. 

My goal was to remove jQuery from the core library (all the js files in the root directory) and make the vanilla version usable without jQuery.

**Comment by 6pac** [comment]
@MarkoBL re return values, ouch, those were already pretty inconsistent.  I think we're best off with a more consistent return type, either a single element or an array depending on the circumstances. 

Since it's going to be breaking anyway, I don't think it matters what we end up with as long as it's (1) useful, (2) as consistent as possible and (3) well documented.

**Comment by ghiscoding** [comment]
we'll of course need a migration Wiki like I've done with [Migration 3.0](https://github.com/6pac/SlickGrid/wiki/Major-version-3.0----Removal-of-jQueryUI-requirement-(replaced-by-SortableJS)). 

I would probably keep the same return, a single element or an array depending on the circumstance, that seems fine by me too

**Comment by MarkoBL** [comment]
> I would probably keep the same return, a single element or an array depending on the circumstance, that seems fine by me too. 
Okay, in this case, I don't have to change anything.

> @MarkoBL What I'm more concerned about is the new `getReturnValues()` that you added to all the `notify`, I posted a comment about that, I'm not sure why it has to change but if it does have to change I would rather have a returned object that includes both the event and the returned value `{ event, value }` instead of having to call an extra `getReturnValues()`. I could be missing the point, so perhaps you could provide us more info on the subject

Well, I don't know anymore, why I made it a function and it doesn't make any sense. I can change it to a property called `returnValue`.

The reason why I return the EventData: The notify stuff was basically broken. Let's say you have two event handlers attached to an event: The first would return `true` and the next `undefined`, notifiy would always return `undefined`, because it always overwrites the older return values and takes the last. And as the grid uses event handlers internally to, this could lead to strange bugs, if external code subscribes to an important internal event.

What I'm doing now: I capture all the the return values and store them in an array. And the first non-`undefined` value is used as the default return value, returned by `getReturnValue()` and if you want to access all returned values from the event handlers, you call `getReturnedValues()`. But as I said, I could change these to properties, instead of functions.

**Comment by ghiscoding** [comment]
The biggest problem that I have with this PR is all around the return value/event from the events, for example the `onBeforeEditCell` is used by every users of the grid and is expecting that if we're passing `false` to the event, then we assume that the Editor will not show up. Will that behavior change, are we going to have to pass an object instead of a boolean as argument? I'm not exactly clear on this part and the behavior did change, we'll have a lot backlash from the community (me included, I still want to pass a boolean). I don't see any examples using the `onBeforeEditCell` event but still, we need to be sure of its usage with this PR. Thanks

https://github.com/6pac/SlickGrid/blob/5367e73a30878513c126a5b6dbf35fe462f7b197/slick.grid.js#LL5341C7-L5344C8

**Comment by 6pac** [comment]
Will try to have a look at these issues on the weekend. The notification comments are troubling: I didn't realise this was broken. Fixing it should be an option rather than working around the whole concept.

From my point of view, there should be no changes to method signatures (ie. input parameters) or return values (outputs), other than that if an parameter or return value (or element of an object as a return value) is a jQuery object or array of objects, it should be changed to the nearest equivalent non-jQuery object or array or objects.

File: slick.grid.js [hunk 19]
```diff
@@ -1973,133 +2027,119 @@ if (typeof Slick === "undefined") {
...
-      // so for equivalent functionality, prior to 1.8 use .width, and after use .outerWidth
-      var verArray = $.fn.jquery.split('.');
-      jQueryNewWidthBehaviour = (verArray[0]==1 && verArray[1]>=8) ||  verArray[0] >=2;
+      const h = ["borderLeftWidth", "borderRightWidth", "paddingLeft", "paddingRight"];
+      const v = ["borderTopWidth", "borderBottomWidth", "paddingTop", "paddingBottom"];
+      const header = _headers[0];
 
-      el = $("<div class='ui-state-default slick-header-column' style='visibility:hidden'>-</div>").appendTo($headers);
       headerColumnWidthDiff = headerColumnHeightDiff = 0;
-      if (el.css("box-sizing") != "border-box" && el.css("-moz-box-sizing") != "border-box" && el.css("-webkit-box-sizing") != "border-box") {
```
Comment (ghiscoding) [inline slick.grid.js:2075]: the `if` condition is missing and is causing some styling issues on my side, I believe it should be
```ts
if (style["box-sizing"] != "border-box" && style["-moz-box-sizing"] != "border-box" && style["-webkit-box-sizing"] != "border-box") {
  //...
```

File: slick.grid.js [hunk 20]
```diff
@@ -1973,133 +2027,119 @@ if (typeof Slick === "undefined") {
...
+      });
+      v.forEach(function (val) {
+        headerColumnHeightDiff += utils.toFloat(style[val]);
+      })
       el.remove();
 
-      var r = $("<div class='slick-row' />").appendTo($canvas);
-      el = $("<div class='slick-cell' id='' style='visibility:hidden'>-</div>").appendTo(r);
-      cellWidthDiff = cellHeightDiff = 0;
-      if (el.css("box-sizing") != "border-box" && el.css("-moz-box-sizing") != "border-box" && el.css("-webkit-box-sizing") != "border-box") {
```
Comment (ghiscoding) [inline slick.grid.js:2088]: same as previous comment, the `if` condition is also missing

**Comment by ghiscoding** [comment]
@MarkoBL I'm starting to look at this, I'm starting to leave some comments on what I see missing and also so far I see that the Column Picker is not showing up at the correct x/y positions

I also checked the `onBeforeEditCell` and that is working correctly, so no issues there

image

File: slick.core.js [hunk 21]
```diff
@@ -675,10 +726,251 @@
...
+				matches.push( elems[ i ] );
+			}
+		}
+
+		return matches;
+	}
+
+  function offset(el) {
+
+    box = el.getBoundingClientRect();
```
Comment (ghiscoding) [inline slick.core.js:752]: `box` is missing the variable declaration `let` or `const` and because it is missing, it can throw this error that I received

```sh
referenceError: box is not defined
 ```

File: slick.core.js [hunk 22]
```diff
@@ -675,10 +726,251 @@
...
+    docElem = document.documentElement;
```
Comment (ghiscoding) [inline slick.core.js:753]: same thing for `docElem`, it is missing the variable declaration and throws an error

File: slick.grid.js [hunk 23]
```diff
@@ -558,69 +572,59 @@ if (typeof Slick === "undefined") {
...
-        }
-
-        $focusSink.add($focusSink2)
-          .on("keydown", handleKeyDown);
-        $canvas
-          .on("keydown", handleKeyDown)
-          .on("click", handleClick)
-          .on("dblclick", handleDblClick)
-          .on("contextmenu", handleContextMenu)            
-          .on("mouseenter", ".slick-cell", handleMouseEnter)
```
Comment (ghiscoding) [inline slick.grid.js:611]: why was this changed from `mouseenter` to `mouseover`? It's not the same event, so why the change?

File: slick.grid.js [hunk 24]
```diff
@@ -1642,60 +1685,71 @@ if (typeof Slick === "undefined") {
...
-      columnElements.each(function (i, colElm) {
+      if (firstResizable === undefined)
+        return;
+
+      for (let i = 0; i < children.length; i++) {
+        const colElm = children[i];
+
         if (i >= columns.length) { return; }
         if (i < firstResizable || (options.forceFitColumns && i >= lastResizable)) {
           return;
```
Comment (ghiscoding) [inline slick.grid.js:1728]: since this `return` was previously used within a jQuery `.each()` method, this `return` will not work inside a `querySelectorAll` loop because it will stop the loop too early. Basically we need to change the `return` to a `continue` for this to work the same as a jQuery each loop

File: slick.core.js [hunk 25]
```diff
@@ -675,10 +726,251 @@
...
+    }
+    Ctor = hasOwn.call( proto, "constructor" ) && proto.constructor;
+    return typeof Ctor === "function" && fnToString.call( Ctor ) === ObjectFunctionString;
+  }
+  function extend() {
+    var options, name, src, copy, copyIsArray, clone,
+      target = arguments[ 0 ],
+      i = 1,
+      length = arguments.length,
+      deep = true;
```
Comment (ghiscoding) [inline slick.core.js:927]: the `deep` default property is wrong, it should be set to `false` since that would follow `jQuery.extend()` more closely. I found this to be a problem when in my SlickGrid libs column filters were not showing up when it should. Basically, the following code in SlickGrid
```js
var m = columns[i] = utils.extend({}, columnDefaults, columns[i]);
```
is supposed to be a shallow copy (work as a pointer), however because of this `deep = true`, it actually became a deep copy and is causing issues in my lib because my lib assume these column copies are shallow copy and not deep copy. After changing default to `deep = false`, that would also follow more closely the NodeJS implementation in external 3rd party lib [node.extend](https://github.com/dreamerslab/node.extend)

https://github.com/dreamerslab/node.extend/blob/62adee3327630c83f3d39395f7e275bfba281ecc/lib/extend.js#L47-L55

```js
module.exports = function extend() {
  var target = arguments[0] || {};
  var i = 1;
  var length = arguments.length;
  var deep = false;
  var options, name, src, copy, copyIsArray, clone;
  // ....
```

File: slick.grid.js [hunk 26]
```diff
@@ -1642,60 +1685,71 @@ if (typeof Slick === "undefined") {
    		return impactedColumns;
     }
 
+    function handleResizeableHandleDoubleClick(evt) {
+      const triggeredByColumn = evt.target.parentElement.id.replace(uid, "");
+      trigger(self.onColumnsResizeDblClick, { triggeredByColumn: triggeredByColumn.getReturnValue() });
```
Comment (ghiscoding) [inline slick.grid.js:1690]: this line throws an error because `const triggeredByColumn` is a string and does not have a `.getReturnValue()` function and it work as expected when triggering without it like so `{ triggeredByColumn: triggeredByColumn }`

**Comment by ghiscoding** [comment]
another issue that I found, there are 2 references of `$.grep` (at line 1334 and 1345) that should be changed to `Slick.Utils.grep`. Also a bigger problem, the Slick.Resizer plugin still uses jQuery and must be converted to vanilla, that shouldn't be too hard since I have nearly the same resizer service in my own lib and I did convert it to vanilla. 

... so I believe that will all of these code change request, it covers it all. I did manage to remove jQuery entirely in my lib and all my E2E tests are now passing after fixing all of the code change requested above. This is getting even closer to be a reality :)

@MarkoBL, if we're not getting news from your side in the coming week, we'll probably merge the PR as it is and create a 2nd PR to fix all the code change I requested. Obviously I would prefer if you were able to fix these chnages, yourself :)

**Comment by ghiscoding** [comment]
@6pac @MarkoBL 
Alright guys, I just finished my entire testing in my Slickgrid-Universal lib, I got all my unit tests and E2E tests working and while fixing them, I found a few issues in this PR and I highlighted these problems as comments in here but since @MarkoBL doesn't seem to be available, I'll go ahead and merge this PR into the `next` branch and create a 2nd PR to fix all these issues comments that I highlighted. I'm proceeding this way so that we can keep all of @MarkoBL great contributions.

@6pac 
Please note that this PR only migrated the files in the root (`slick.grid`, `slick.core`, `slick.dataview`, ...), however **it does not** include any Controls & Plugins. However, I don't think we can ship a new `major` 4.0 version without rewriting and migrating all Controls & Plugins to vanilla JS... sooo it will still need more work. 

On the other hand, I was able to completely remove jQuery in Slickgrid-Universal, which is why I found a few issues in here, because I already rewrote all of these Controls & Plugins to vanilla JS (actually in TypeScript for my lib), so I should be able to port the code back into here and hopefully that shouldn't be too too long. I would really like to be ready for a release at the start of the summer (North summer that is 😉)

@6pac also note that your other PR should be rebased with the `next` branch since your code will be shipped at the same time under this next major branch

**Comment by 6pac** [comment]
great, thanks for all your work. i'm still extremely busy (probably till July) so won't have a lot of time to dedicate to this. Will definitely rebase that PR though, and can have a look at small issues.
